### DISCUSSION TITLE
Running test with vitest and @vue/test-utils

### DISCUSSION THREAD

**Discussion by DesKevinMendez** [nuxt/nuxt#15908] [General]
Report a bug.

I have a set of test with `vitest` and `@vue/test-utils` but, currently I receive the next error.

Note: is a new project created by `npx nuxi init nuxt-app`

<img width="1349" alt="Captura de Pantalla 2022-06-01 a la(s) 14 48 49" src="https://user-images.githubusercontent.com/32910223/171498851-dee3a961-35af-49b8-bcab-2547ee68b908.png">

My package.json is

```
{
  "private": true,
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "test:unit": "yarn vitest run --dir test",
    "preview": "nuxt preview",
    "lint": "eslint --ext .js,.ts,.vue --ignore-path .eslintignore ."
  },
  "devDependencies": {
    "@nuxt/test-utils": "^3.0.0-rc.3",
    "@nuxt/test-utils-edge": "^3.0.0-rc.3-27567768.c1f728e",
    "@nuxtjs/eslint-config-typescript": "^10.0.0",
    "@typescript-eslint/eslint-plugin": "^5.27.0",
    "@typescript-eslint/parser": "^5.27.0",
    "@vue/eslint-config-standard": "^7.0.0",
    "@vue/test-utils": "^2.0.0-rc.18",
    "c8": "^7.11.3",
    "commitizen": "^4.2.4",
    "eslint": "^8.16.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-nuxt": "^3.2.0",
    "eslint-plugin-prettier": "^4.0.0",
    "eslint-plugin-vue": "^9.1.0",
    "git-cz": "^4.9.0",
    "happy-dom": "^5.0.0",
    "nuxt": "3.0.0-rc.3",
    "vite-plugin-eslint": "^1.6.1",
    "vitest": "^0.13.1"
  },
  "config": {
    "commitizen": {
      "path": "git-cz"
    }
  },
  "dependencies": {}
}

```

And my `nuxt.config.ts`

```
import { defineNuxtConfig } from 'nuxt'
import eslintPlugin from 'vite-plugin-eslint'

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  vite: {
    plugins: [
      eslintPlugin({
        emitWarning: true,
        emitError: true
      })
    ]
  }
})
```

And my test:

```
import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import App from '@/app.vue'

describe('app.vue', () => {
  it('component must be mounted correctly', () => {
    const wrapper = mount(App)
    expect(wrapper).toBeTruthy()
  })
})
```

I don´t have idea what is wrong with that. Am I missing a configuration or am I need install a new package?

**Comment by itsmnthn** [comment]
Seems Path related issue

Using nuxt's ts config's paths as alias in vitest config solves the issue

```TypeScript
// file alias.ts

import { resolve } from 'path'

const r = (p: string) => resolve(__dirname, p)

export const alias: Record<string, string> = {
  '~~': r('.'),
  '~~/': r('./'),
  '@@': r('.'),
  '@@/': r('./'),

  ... other paths
}
```

```TypeScript
// vitest.config.ts

import { defineConfig } from 'vite'
import { alias } from './alias'

export default defineConfig({
  root: '.',
  esbuild: {
    tsconfigRaw: '{}',
  },
  resolve: {
    alias,
  },
})

```

**Reply by Maxim-Filimonov** [reply]
What is the second file? Is it `vite.config.js`? I tried to create `vite.config.js` it with this content, and it didn't work.
Figured it out myself.
Use in `vitest.config.ts`:
```
import { defineConfig } from "vitest/config";
import { alias } from "./alias";

export default defineConfig({
  resolve: {
    alias,
  },
  test: {
  // rest of vitest config 
  }
})
  ```

**Comment by DesKevinMendez** [comment]
Fix it. Thanks @itsmnthn

**Comment by DesKevinMendez** [comment]
I wrote my vitest.config.ts

```
import { resolve } from 'path'
import fs from 'fs'
import vue from '@vitejs/plugin-vue'
import { defineConfig } from 'vitest/config'

const NuxtTsConfig = (fs.readFileSync('./.nuxt/tsconfig.json')).toString()
const tsConfigFormated = JSON.parse(NuxtTsConfig
  .replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? '' : m))

const r = (p: string) => resolve(__dirname, p)

export const alias: Record<string, string> = {}

Object.entries(tsConfigFormated.compilerOptions.paths)
  .forEach(([key, value]) => {
    alias[key] = r(value[0])
  })

export default defineConfig({
  root: '.',
  plugins: [
    vue()
  ],
  resolve: {
    alias
  },
  test: {
    environment: 'jsdom'
  }
})

```

**Comment by Arkadipta1998** [comment]
@DesKevinMendez @itsmnthn How did you handle the nuxt auto imports in vitest? It says 'reactive is not defined'. In Nuxt 3 all the auto imports are exposed via .nuxt/imports so I included that file in setupFiles under test like so -

````
// vitest.config.ts

export default defineConfig({
  root: '.',
  esbuild: {
    tsconfigRaw: '{}',
  },
  plugins: [
    Vue(),
  ],
  test: {
    globals: true,
    environment: 'jsdom',
    deps: {
      inline: [/@nuxt\/test-utils-edge/],
    },
    setupFiles: ['./.nuxt/imports.d.ts']
  },
  resolve: {
    alias
  }
})
````

But this approach gave me an error like this - Package import specifier "#build/app.config.mjs" is not defined in package
My nuxt version `3.0.0-rc.10`

**Reply by nikitin-dev** [reply]
Same issue

**Reply by BIGScream** [reply]
same....

**Comment by yassilah** [comment]
So i've been having the exact same issue as everyone, I ended up making a Vite plugin that resolves paths and adds auto-imports; hope that's useful:

```ts
// vitest.config.ts
import { defineConfig } from 'vitest/config'
import { join, resolve } from 'path'
import AutoImport from 'unimport/unplugin'
import Vue from '@vitejs/plugin-vue'

export default defineConfig({
    plugins: [
        Vue(),
        AutoImport.vite({
            presets: [
                {
                    package: getBuildDir('imports.d.ts')
                }
            ]
        }),
        {
            name: 'nuxt:vitest',
            async config() {
                return {
                    resolve: {
                        alias: pathsToAlias(await getTsConfigPaths())
                    },
                    test: {
                        environment: 'happy-dom',
                        deps: {
                            inline: ['nuxt/dist']
                        }
                    }
                }
            }
        }
    ]
})

function getBuildDir(path: string) {
    // Unfortunaely, vite.config.ts does not support top-level await,
    // so we can't actually resolve the buildDir from Nuxt as follows:

    //     const { loadNuxt } = await import('nuxt')
    //     const nuxt = await loadNuxt({})
    //     return join(nuxt.options.buildDir, path)

    return join(__dirname, '.nuxt', path)
}

/**
 * Turn tsconfig paths into aliases.
 */
function pathsToAlias(paths: Record<string, string[]>) {
    return Object.fromEntries(
        Object.entries(paths).map(([key, value]) => [
            key,
            resolve(__dirname, value[0])
        ])
    )
}

/**
 * Get the tsconfig from Nuxt.
 */
async function getTsConfig() {
    const { readTSConfig } = await import('pkg-types')
    return readTSConfig(getBuildDir('tsconfig.json'))
}

/**
 * Get the paths from Nuxt's tsconfig.
 */
async function getTsConfigPaths() {
    const { compilerOptions } = await getTsConfig()
    return compilerOptions?.paths as Record<string, string[]>
}
```

**Comment by yassilah** [comment]
Ok so I just made that into a simple plugin: https://github.com/yassilah/vite-plugin-nuxt-test

```ts
// vitest.config.ts
import { defineConfig } from 'vitest/config'
import NuxtVitest from 'vite-plugin-nuxt-test'

export default defineConfig({
    plugins: [NuxtVitest()],
    test: {
        // Your own configuration
    }
})
```

**Reply by MrKoopie** [reply]
Thanks for the great work. I have some issues with this plugin. If you check out https://stackblitz.com/edit/github-rdkvdc-eycaii?file=README.md&view=editor, you will see that running yarn vitest results in:

```
[Vue warn]: Failed to resolve component: WidgetCard
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement. 
```

Can you confirm if your plugin should resolve the autoloading issues of components?

**Reply by MrKoopie** [reply]
Furthermore, in https://stackblitz.com/edit/github-rdkvdc-br4pek?file=components/Info.vue I have used #imports. This also does not seem to resolve the issue.

**Reply by larsalbrecht** [reply]
Did not help me.
Before*:
```Package import specifier "#build/app.config.mjs"```

\* With alias and setupFiles


With your plugin:
```"Error: Cannot find module /src/node_modules/@pinia/nuxt/node_modules/pinia/dist/pinia"```

If I still add setupFiles (['./.nuxt/imports.d.ts', './.nuxt/nuxt.d.ts']):
```"Error: [vite-node] Failed to load /src/node_modules/nuxt/dist/head/runtime"```

**Reply by MrKoopie** [reply]
I had a similar issue with Pinia and it was fixed by updating the Nuxt Pinia library. However, I still had other issues.

**Reply by mcauto** [reply]
@yassilah 
I get error below.
could you help me?
I test a composables.

```
Error: nuxt instance unavailable
chunk-node-pkg.e7e5686f.js:7374
 ❯ Module.useNuxtApp node_modules/nuxt/dist/app/nuxt.mjs:165:13
chunk-node-pkg.e7e5686f.js:7374
    164|   Object.defineProperty(obj, key, { get: () => val });
    165| }
    166| export function defineAppConfig(config) {
       |           ^
    167|   return config;
    168| }
```

**Reply by yassilah** [reply]
Hey, sorry for the late reply. I'm not sure what your issue is @mcauto but FYI @danielroe has published a package that lets you run tests in your Nuxt environment and should be a lot more reliable than my module: https://github.com/danielroe/nuxt-vitest 😊

**Comment by anayatkhan1** [comment]
# Help me to solve this testing error.🙏
## For this Feedback form component, I'm using the NUXT UI module.

`<template>
    <div class="grid grid-cols-6 rounded-3xl mt-16">
        <div class="w-full max-w-sm col-start-2 col-span-4 lg:mx-auto">
            <UCard>
                <template #header
                    ><h2 class="text-3xl font-bold tracking-tight text-center">
                        Feedback
                    </h2></template
                >`

                <form @submit.prevent="submitForm">
                    <div class="grid grid-cols-2 gap-y-6">
                        <div class="col-span-2">
                            <UFormGroup label="Name">
                                <UInput
                                    name="user-name"
                                    type="text"
                                    size="lg"
                                    placeholder="Your name"
                                    icon="i-heroicons-user-20-solid"
                                    
                            /></UFormGroup>
                        </div>
                        <div class="col-span-2">
                            <UFormGroup label="Email">
                                <UInput
                                    name="user-email"
                                    type="email"
                                    size="lg"
                                    placeholder="you@example.com"
                                    icon="i-heroicons-envelope"
                                    
                            /></UFormGroup>
                        </div>
                        <div class="col-span-2">
                            <UFormGroup label="Feedback">
                                <UTextarea
                                    name="user-message"
                                    size="lg"
                                    :rows="5"
                                    placeholder="Your message"
                                    
                            /></UFormGroup>
                        </div>
                        <div class="col-span-2">
                            <UButton
                                name="user-send-message"
                                block
                                size="lg"
                                type="submit"
                                label="Send message"
                            />
                        </div>
                    </div>
                </form>
            </UCard>
        </div>
    </div>
</template>
<script setup lang="ts">
import { z } from "zod";
const toast = useToast();
const runtimeConfig = useRuntimeConfig();

interface Feedback {
    name: string;
    email: string;
    feedback: string;
}
const formData = ref<Feedback>({
    name: "",
    email: "",
    feedback: "",
});
const validationSchema = z.object({
    name: z.string().nonempty("Name is required"),
    email: z
        .string()
        .nonempty("Email is required")
        .email({ message: "Must be a valid email" }),
    feedback: z.string().nonempty("Feedback Message is required"),
});

const submitForm = async () => {
    const valid = await validationSchema.safeParseAsync(formData.value);
    if (!valid.success) {
        const message = valid.error.issues[0].message;
        toast.add({
            title: message,
            color: "red",
        });
        return;
    }
    try {
        const { data, pending, error, refresh } = await useFetch(
            "/site/feedback",
            {
                method: "POST",
                baseURL: String(runtimeConfig.public.BASE_API_URL),
                body: JSON.stringify(formData.value),
            }
        );
        toast.add({
            title: "Thank you for your feedback!",
            color: "green",
        });
    } catch (error: any) {
        toast.add({
            title: "Oops! There was an error",
            description: error.value.message,
            color: "red",
        });
    }
};
</script>
# This is vitest code for testing this form component

import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import Feedback from '../../pages/site/feedback.vue';

describe('Feedback Form is Testing', () => {
    
    it('components should be exists & Snapshot should be Matched', async () => {
        const wrapper = mount(Feedback, {
            
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Elements should be exists', async () => {
        const wrapper = mount(Feedback);

        const Input =  await wrapper.find('UInput');

        expect(Input.exists());
    });

});


![unit 1](https://github.com/nuxt/nuxt/assets/73161735/d41618a5-a05e-438e-8cfa-0e09cc2ead67)

#### I have been attempting to resolve this issue by seeking assistance from both **Google** and a **Discord** server { Nuxt, Vue Land, Vitest Dev,  }. Despite my efforts, I have not yet found a solution. I kindly request your help in addressing this matter.

Thank You ⭐
